src/middlewares.rs

```diff
@@ -18,30 +21,66 @@ pub struct UserState {
 
 pub type CookieUserState = Extension<UserState>;
 
+impl<'a> From<UserState> for Cookie<'a> {
+    fn from(userstate: UserState) -> Self {
+        let userstate_json =
+            serde_json::to_string(&userstate).expect("failed to serialize to json");
+
+        let userstate_base64 = STANDARD_NO_PAD.encode(userstate_json);
+
+        Cookie::build(("state", userstate_base64))
+            .http_only(true)
+            .secure(true)
+            // .max_age(Duration::MAX) // do we want exp date?
+            // .domain(value) // TODO: set domain?
+            .build()
+    }
+}
+
+impl Default for UserState {
+    fn default() -> Self {
+        let selection = Vec::new();
+        UserState { selection }
+    }
+}
+
+impl<'a> TryFrom<Cookie<'a>> for UserState {
+    type Error = AppError;
+
+    fn try_from(cookie: Cookie<'a>) -> Result<Self, Self::Error> {
+        let cookie_base64 = cookie.value();
+
+        let cookie_json = STANDARD_NO_PAD.decode(cookie_base64).map_err(|err| {
+            error!("invalid base64 encoded cookie: {}", err);
+            StatusCode::BAD_REQUEST
+        })?;
+
+        let cookie_json = str::from_utf8(cookie_json.as_ref()).map_err(|err| {
+            error!("invalid utf 8 string: {}", err);
+            StatusCode::BAD_REQUEST
+        })?;
+
+        let userstate: UserState = serde_json::from_str(cookie_json).map_err(|err| {
+            error!("invalid json encoded cookie: {}", err);
+            StatusCode::BAD_REQUEST
+        })?;
+
+        Ok(userstate)
+    }
+}
+
 pub async fn parse_cookie(
     cookie: CookieJar,
     mut req: Request,
     next: Next,
 ) -> Result<Response, (StatusCode, String)> {
```

Comment:
This should not return a status code because it does not live inside of a request handler. What if I want to use this code somewhere else? What if there is a way to recover gracefully from a cookie parse failure?

Reply:
this is a middleware and shouldn't be called outside the context of it being a middleware.

I'm returning a status code since (imo) any ops after that failure shouldn't be allowed, since it seems like a lot of our actions rely on the fact that the user state is valid

Reply:
Does axum expect the middleware to return a status code and will it short-circuit in that case? If so, please link the relevant documentation showing that dealing with status codes in middleware is a blessed path.

Reply:
sure, this is the example I based my code off.
https://docs.rs/axum/latest/axum/middleware/index.html#passing-state-from-middleware-to-handlers

Reply:
Ok, I think I understand. If the extension returns an `Err` then axum will never call the request handler.